# Linear Regression

In a linear regression model, we intend to approach problems where the response variable is quantitative. The relationship is expressed in the form of an equation or a model connecting the response or dependent variable and one or more explanatory or predictor variable. The general mathematical equation for linear regression is given by -

$$
\begin{align*}
Y &= f(X_1, X_2, ..., X_p) + \epsilon \\
Y &= \beta_0 + \beta_1 X_1 + \beta_2 X_2 + ... + \beta_p X_p + \epsilon
\end{align*}
$$

A reasonable approach to fit a line (find the $\beta$ values) is to minimize the amount by which the fitted value differs from the actual value. An assumption in this approach is that the predictor variables are observed without any uncertainty but the response variable may have variance due to irreducible error. This leads to the *Ordinary Least Squares* method.

Alternate fitting method: Another approach for the fit is to assume that both the independent and dependent variables have uncertainty in measurement. This leads to the *Total Least Squares* method.

**Ordinary Least Squares** <br>
This approach is an MLE solution for obtaining the best regression line. This method minimizes the *Residual Sum of Squares (RSS)* or the *Mean Squared Error (MSE)*

$$ RSS = \sum_{i=1}^{n} (y_i - \hat{y}_i)^2 $$

$$ MSE = \frac{1}{n} \sum_{i=1}^{n} (y_i - \hat{y}_i)^2$$

Therefore the objective of the approach is given as,

$$
\min_{\beta_0, ..., \beta_p} \frac{1}{n} \sum_{i=1}^{n} (y_i - \hat{y}_i)^2 \: \text{, or} \\
\\
\min_{\beta_0, ..., \beta_p} \frac{1}{n} \sum_{i=1}^{n} (y_i - [\hat{\beta_0} + \hat{\beta_1} x_1 + \hat{\beta_2} x_2 + ... + \hat{\beta_p} x_p])^2 \\
$$

Mathematically, for simple linear regression the $\beta$ values are given as follows.

$$\hat{\beta}_1 = \frac{\sum (x_i - \bar{x})(y_i - \bar{y})}{\sum (x_i - \bar{x})^2} = r_{xy} \frac{\sigma_y}{\sigma_x}$$

$$\hat{\beta}_0 = \bar{y} - \hat{\beta_1}\bar{x}$$

Note:
* We can see that when $\hat{\beta}_1 = 0$, then the $\hat{\beta}_0$ estimate is just the mean of the response variable and this results in a model called the *null model*
* Here $r_{xy}$ represents the correlation between x and y for the data (sample). Pearson correlation uses $\rho_{xy}$ for population and $r_{xy}$ for samples.
* When using the lag of response itself as the predictor (Autoregression), we can see that the $\hat{\beta}_1$ estimate reduces to just the $r_{xy}$ as the variance of $x$ and $y$ are (approximately) the same. Therefore, in autoregression, the beta corresponds to the value of auto-correlation between the response and the lagged response.

In [1]:
import pandas as pd
import statsmodels.api as sm
from sklearn.datasets import load_diabetes

In [2]:
data = load_diabetes()

X = pd.DataFrame(data.data, columns=data.feature_names)
X = sm.add_constant(X[['age', 'sex', 'bmi', 'bp']])
y = data.target

In [3]:
model = sm.OLS(y, X)
results = model.fit()

display(results.summary())

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.400
Model:                            OLS   Adj. R-squared:                  0.395
Method:                 Least Squares   F-statistic:                     72.91
Date:                Thu, 20 Jul 2023   Prob (F-statistic):           2.70e-47
Time:                        03:56:31   Log-Likelihood:                -2434.2
No. Observations:                 442   AIC:                             4878.
Df Residuals:                     437   BIC:                             4899.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        152.1335      2.853     53.329      0.000     146.527     157.740
age           37.2406     64.117      0.581      0.562     -88.776     163.258
sex         -106.5762     62.125     -1.716      0.087    -228.677      15.525
bmi          787.1817     65.424     12.032      0.000     658.597     915.766
bp           416.6725     69.495      5.996      0.000     280.087     553.258
==============================================================================
Omnibus:                        9.858   Durbin-Watson:                   1.933
Prob(Omnibus):                  0.007   Jarque-Bera (JB):                6.464
Skew:                           0.146   Prob(JB):                       0.0395
Kurtosis:                       2.485   Cond. No.                         28.4
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## Coefficients

Assume the above example. Considering the independent variables `age`, `sex`, `bmi`, `bp` to predict the `disease progression` as a regression response. Mathematically - 

$$ \hat{Prog} = \hat{\beta}_0 + \hat{\beta}_1.age + \hat{\beta}_2.sex + \hat{\beta}_3.bmi + \hat{\beta}_4.bp + \epsilon $$


**Interpretation** <br>
Each coefficient can be interpreted as the marginal effect on the response variable with respect to the independent variable considering all other independent variables are held constant. Mathematically,

$$\frac{\partial y}{\partial x_i} = \beta_i$$

We can generalize the interpretation of a coefficients in relation to the marginal effect using the following table.


| Equation | Marginal Effect |
|---|---|
| $y = b_0 + b_1 x$ | $\frac{\partial y}{\partial x} = b_1$ |
| $y = b_0 + b_1 ln(x)$ | $\frac{\partial y}{\partial x} = \frac{b_1}{x}$ |
| $ln(y) = b_0 + b_1 x$ | $\frac{\partial y}{\partial x} = b_1 y$ |
| $ln(y) = b_0 + b_1 ln(x)$ | $\frac{\partial y}{\partial x} = b_1 \frac{y}{x}$ |


**Standard Error** <br>
Here, since we're considering only a sample of the population as our dataset, the $\beta$ etimates can be an over-estimate or under-estimate. But the OLS method produces an unbiased estimate of the $\beta$, meaning the average over a large number of estimates converges to the true value of the $\beta$. Therefore, for the dataset (sample), we can estimate the *Standard Error (SE)* similar to estimating the SE of the mean of a dataset.

In our example, the SE for the estimate of the coefficient for `age` ($\hat{\beta}_1$) is $64.117$.

**T-Statistic** <br>
This value by itself is not useful unless we can compare it with the magnitude of the coefficient estimate. For age, we have $\hat{\beta}_1 = 37.2406$ which means that the coefficient may differ from the true value by a value of $64.117$ on average. We can perform hypothesis testing to make sure the coefficient is not zero.
* $H_0$ - There is no relationship between $X$ and $Y$, i.e. $\beta = 0$
* $H_a$ - There is some relationship between $X$ and $Y$, i.e. $\beta \neq 0$

The t-statistic is calculated for a degree of freedom of $n-2$. Mathematically,

$$t = \frac{\hat{\beta}-0}{SE(\hat{\beta})}$$

**P-value** <br>
Given the t-statistic, we can find the p-value as the probability of observing a value $|t|$ or larger given $\beta = 0$. In other words, the p-value is a probability value that quantifies the likelihood of obtaining the observed results (or more extreme) under the assumption that the null hypothesis is true.

## Model Assessment

**Residual Standard Error (RSE)** <br>
This is the estimate of the quantity $\sigma$ where $\sigma^2$ is then variance of $\epsilon$. The RSE is calculated as - 

$$RSE = \sqrt{\frac{1}{n-p-1} RSS} = \sqrt{\frac{1}{n-p-1} \sum_{i=1}^{n} (y_i - \hat{y_i})^2}$$

RSE is measured in the same units as the response variable, and the value of RSE roughly corresponds to the average amount the response will deviate from the true regression line.

Note: Root Mean Squared Error (RMSE) is a similar measure, the difference being the mean considers $n$ in the denominator instead of the *degrees of freedom* of the residuals.

$$RMSE = \sqrt{\frac{1}{n} \sum_{i=1}^{n} (y_i - \hat{y_i})^2}$$


**R-squared** <br>
R-squared measures the proportion of variability in $Y$ that can be explained using $X$. Mathematically - 

$$R^2 = \frac{TSS-RSS}{TSS} = 1 - \frac{RSS}{TSS} $$

where,
* $RSS = \sum (y_i - \hat{y}_i)^2$. RSS represents the sum of squares of the residuals of the regression model.
* $TSS = \sum (y_i - \bar{y})^2$. TSS represents teh sum of squares of the residuals of the null model.

NOTE: For simple linear regression (including Autoregression with one predictor), the $R^2$ statistic is the equal to the squared correlation between the response and the predictor, $r^2$


**F-Statistic** <br>
F-statistic is a metric used to perform hypothesis testing to whether all the coefficients are 0. 
* $H_0$ - $\beta_1 = \beta_2 = ... = \beta_p = 0$
* $H_a$ - At least one $\beta_i$ is non-zero

$$F = \frac{(TSS-RSS)/p}{RSS/(n-p-1)}$$

F-statistic can easily be converted to a p-value to understand more easily the significance of the F-statistic's value given the degrees of freedoms for the statistic. Use the F-distribution with $df_1 = p$ and $df_2 = n-p-1$ to determine the p-value.

NOTE: It is important to evaluate the F-statistic and the overall p-value. Consider a large number of independent variables, say $p=100$. If we have a significance level of 0.05, 5 variables may have a significant p-value just by chance. However, the F-statistic does not suffer from this problem because it adjusts for the number of predictors.

> AIC, BIC, Adjusted R<sup>2</sup> - can be used for model/variable selection

## Variable Selection - Classical Approaches

The following variable selection methods can be used when it is cumbersome to build $2^p$ models and evaluate each model to choose the best subset of variables for the linear regression model.

1. **Forward selection** - Start with the null model, and perform a greedy process of adding one predictor at a time (with the lowest p-value). This procedure is repeated until a stopping criteria is reached.

2. **Backward selection** - Start with all variables in the model, and perform a greedy process of removing one predictor at a time (with the highest p-value). This procedure is repeated until a stopping criteria is reached.

3. **Mixed selection** - This is a combination of forward and backward selection. Start with no variables in the model and follow forward selection. When a variable crosses a significance threshold in the p-value, start removing the variable.

## Predictions

1. Confidence interval: Reducible Error -->  how close $\hat{Y}$ will be to $f(X)$
2. Prediction interval: Irreducible Error + Reducible Error -->  how close $\hat{Y}$ will be to $Y$

> Interaction terms
> 
> Power terms
> 
> Potential Problems
> 1. Non-linearity of the response-predictor relationships
> 2. Correlation of error terms
> 3. Non-constant variance of error terms
> 4. Outliers
> 5. High-leverage points
> 6. Collinearity